crates/tracing-instrumentation/src/lib.rs

```diff
@@ -96,31 +99,119 @@ where
+
+    let tracer = SpanModifyingTracer::new(
+        provider
+            .tracer_builder("opentelemetry-otlp")
+            .with_version(env!("CARGO_PKG_VERSION"))
+            .build(),
+    );
+    // let _ = opentelemetry::global::set_tracer_provider(provider);
+
+    Ok(Some(
+        tracing_opentelemetry::layer()
+            .with_location(false)
+            .with_threads(false)
+            .with_tracked_inactivity(false)
+            .with_tracer(tracer)
+            .with_filter(EnvFilter::try_new(&opts.filter)?)
+            .with_filter(filter_fn(|meta| meta.target() == DEPLOYMENT_TARGET))
+            .boxed(),
+    ))
+}
+
+#[allow(clippy::type_complexity)]
+fn build_runtime_tracing_layer<S>(
+    common_opts: &CommonOptions,
+    service_name: String,
+) -> Result<Option<Box<dyn tracing_subscriber::Layer<S> + Send + Sync + 'static>>, Error>
+where
+    S: tracing::Subscriber
+        + for<'span> tracing_subscriber::registry::LookupSpan<'span>
+        + Send
+        + Sync,
+{
+    let opts = match &common_opts.tracing_runtime {
+        Some(opts) => opts,
+        None => return Ok(None),
+    };
+
+    let resource = opentelemetry_sdk::Resource::new(vec![
+        KeyValue::new(
+            opentelemetry_semantic_conventions::resource::SERVICE_NAME,
+            service_name.clone(),
+        ),
+        KeyValue::new(
+            opentelemetry_semantic_conventions::resource::SERVICE_NAMESPACE,
+            "Restate",
+        ),
+        KeyValue::new(
+            opentelemetry_semantic_conventions::resource::SERVICE_INSTANCE_ID,
+            format!("{}/{}", common_opts.cluster_name(), common_opts.node_name()),
+        ),
+        KeyValue::new(
+            opentelemetry_semantic_conventions::resource::SERVICE_VERSION,
+            env!("CARGO_PKG_VERSION"),
+        ),
+    ]);
+
+    // the following logic is based on `opentelemetry_otlp::span::build_batch_with_exporter`
+    // but also injecting ResourceModifyingSpanExporter around the SpanExporter
+
+    let mut tracer_provider_builder = opentelemetry_sdk::trace::TracerProvider::builder()
+        .with_config(opentelemetry_sdk::trace::Config::default().with_resource(resource));
+
+    let header_map = HeaderMap::from_iter(HashMap::from(opts.headers.clone()));
+
+    let exporter = SpanExporterBuilder::from(
+        opentelemetry_otlp::new_exporter()
+            .tonic()
+            .with_endpoint(&opts.endpoint)
+            .with_metadata(MetadataMap::from_headers(header_map)),
+    )
+    .build_span_exporter()?;
+
+    tracer_provider_builder = tracer_provider_builder.with_span_processor(
+        BatchSpanProcessor::builder(exporter, opentelemetry_sdk::runtime::Tokio).build(),
+    );
+
+    if let Some(path) = &opts.json_path {
         let exporter = JaegerJsonExporter::new(
             path.into(),
             "trace".to_string(),
             service_name,
             opentelemetry_sdk::runtime::Tokio,
         );
```

Comment:
I am not sure tbh, but it's up to the operator to set the configuration properly. We can add a check that if both json exporters are set that they can't be the same file.

Reply:
I guess this is mainly a problem because of defaulting to tracing_runtime if tracing_services is not explicitly configured. If the users explicitly configures both tracings to write to the same file, then this is on them, I would say.